### DISCUSSION TITLE
Synth ideas

### DISCUSSION THREAD

**Discussion by maks** [xiphonics/picoTracker-pcb#5] [General]
@democloid I'm very keen to try my hand at building some synths into picoTracker too so I thought I'd jot down some thoughts about this, though I'm sure you've considered most if not all of these already.

In terms of resources, I guess cpu budget constraints are going to be important but i think the biggest limit/headache is going to be the lack of an FPU and vector (SIMD) units in the RP2040, so maths will need to be done in fixed point to avoid the slowness of floating point being done in software.

I'll start with the synth I'm most familiar with which is an emulation of the DX7.  A few years ago I revived the wonderful "music synth for android" (msfa) app from Raph Levien  under the moniker of [DroidSynth](https://github.com/maks/droid-synth) and its core engine "msfa" is the basis for the quite popular Dexed VST plug. More recently I broke out the msfa engine into a separate [CMake based lib](https://github.com/maks/libmsfa) to use in a new v2 of my app this time built in Flutter so that should be pretty easy to plugin to picoTracker and using it would mean getting access to a **huge** library of patches including sounds of half the pop songs and movie soundtracks of my 80s childhood 🙂. On the downside is that Raph spent a lot of effort doing a core in hand written Neon assembly and the plain C version uses FP so the perf on the RP2040 of a naive build will be pretty terrible as evidenced by at least [one person who attempted it](http://www.sayatone.com/fmsynthesis/raspberry-pi-pico-dx7-clone/) and their [comments on the port](https://github.com/asb2m10/dexed/discussions/311). I'm fairly sure they didnt realise that the poor perf is most likely due to teh FP being done in software due to lack of a FPU, but since they never published their code, I can't be sure.

But just looking at doing a FM synth from scratch rather than attempting to emulate on old 80s hardware one should be quite doable, there are already a number of open src FM synths out there and I vaguely remember seeing in the readme of 1 that a single audio frame needed around 250 math ops so that should be quite doable as I think you said @democloid that 1 core is mostly not sued at the moment? 

I've got thoughts about other synths but this is already getting too long, so I'll post them as follow up comments in this thread.

**Comment by democloid** [comment]
Yes, while FP code is not necessarily a dealbreaker, it will, in general, be too expensive to be of any practical use. This is where I most struggled, since it's hard to find fixed point math synth engines.

One core is mostly not used for the moment, but at the same time, I wasn't successful being able to parallelize the operations. The way it currently works is that the UI/inputs runs in core0 and the audio driver in core1. The load on core0 is very low. (as an unrelated side note, this is the reason why battery life would be greatly improved by significantly lowering the clock speed and only turn it up when playing)

Due to memory restrictions, any delay, reverb, etc is probably out of the question.

Some synths I tried:
* reSID: as expected, pretty heavy, but you can run one channel on it
* [crSID](https://csdb.dk/release/?id=219530): This works pretty well and fast enough to play 8 channels on SID synths (I think, never finished the code)
* [sunvox engine](https://github.com/bohwaz/sunvox): While sunvox went closed source a while back, this old version of the engine contains a lot of interesting synths and the ability to compose synths voices in a modular way. Haven't really tested it, but the code is fixed point, so I thought it might be promising.

**Reply by maks** [reply]
Funny you mention sunvox as thats what I used as the [engine for my ML-2 project recently](https://github.com/maks/ml_2) !
NightRadio's an amazing dev and I've been also very impressed with sunvox and his pixilang scripting lang.

But good news there, its not well known that the whole audio engine of the current version of sunvox *is* available under an a MIT license, its just "hidden" away in the source for [pixilang](https://github.com/maks/pixilang/tree/main/lib_sunvox) 🙂  and yes thats a **great** idea! I'm not sure why I hadn't thought of it as it was originally written to run on things as low powered as the old Palm PDAs! 🙂

**Reply by maks** [reply]
@democloid also I was hoping to ask you a couple of questions privately, if you are willing to, could you please email me on maks at manichord.com ?

**Reply by democloid** [reply]
That's super interesting! thou I had a feeling that the underlying engine might not have changed much, but not way to prove of course. I found that the modules are for the most part the same they there are today.

Sending you email

**Comment by maks** [comment]
Sorry I forgot to come back to continue include the remainder of what I had written about other possible synths:

Another interesting synth I'm aware of is [Peter Salomonsen's wasm synth](https://github.com/petersalomonsen/javascriptmusic)   as assemblyscript is easy to read but close enough to C that could be easily ported or even to actual ARM assembly, though again would need to switch from float to fixed point math. The most interesting part of Peters work though is what he's done on "physical modelling" synths which he has [written up and has easy to ready code for](https://petersalomonsen.com/articles/assemblyscriptphysicalmodelingsynthesis/assemblyscriptphysicalmodelingsynthesis.html) though I have a bad feeling that for this the memory requirements may be too high.

Also its probably worth mentioning for any future readers that picoTracker does already have some "synth" capabilities by supporting single cycle waveforms as long as the [wave files are in the required format](https://chipmusic.org/forums/topic/13105/samples-which-wont-work-in-lgpt-and-samples-that-will/). It would probably be possible to follow in Polyend Trackers footsteps here and also support basic granular synth functionality as single  cycle waves and granular are the 2 things they support in their sample-based tracker also for probably the same reason as its jsut a variation of sample playback and easy to do on limited hardware (I think the PE tracker is based on Teensy 3).

**Reply by smutax** [reply]
Hi. This is very nice project. Perhaps you want to have a look at Amy library here
https://github.com/bwhitman/amy
It was done in fixed point and happily runs on Pi Pico. It's being continually improved and got a friendly devs and community. All the very best.

**Reply by maks** [reply]
@smutax  thanks for the pointer to amy, I had not come across it before and it looks very interesting!

**Reply by maks** [reply]
I went looking for what performance is like with Amy and found this on their discord which just blew my mind when it comes to dx7 style FM! And the fact that they also have a python script to convert from DX7 patches is just awesome for me! (as there are something like 60k DX7 patches available online)

I'll certainly be looking into getting Amy running on the pico! Thanks again @smutax 

FYI @democloid :

![image](https://github.com/democloid/picoTracker-pcb/assets/71999/e3baa15d-63dd-4e19-858c-10bc932d1cbe)

**Reply by smutax** [reply]
Haha. You are welcome! I'm watching and sizing up Amy and co. for a while now. It's got massive potential! Keep up a good work and best of luck with a projects you are involved with.
